drivers/gpio/gpio_dw.c

```diff
@@ -177,18 +267,26 @@ static inline void dw_pin_config(struct device *port,
 	const struct gpio_dw_config *config = port->config->config_info;
 	u32_t base_addr = dw_base_to_block_base(context->base_addr);
 	u32_t port_base_addr = context->base_addr;
-	int interrupt_support = dw_interrupt_support(config);
-
-	if (interrupt_support) {
-		/* clear interrupt enable */
-		dw_set_bit(base_addr, INTEN, pin, 0);
+	u32_t dir_port = dw_get_dir_port(port_base_addr);
+	bool pin_is_output, need_debounce;
+
+	/* set direction and init value */
+	pin_is_output = (flags & GPIO_OUTPUT);
+	dw_set_bit(port_base_addr, dir_port, pin, pin_is_output);
+
+	if (pin_is_output) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			gpio_dw_port_set_bits_raw(port, BIT(pin));
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			gpio_dw_port_clear_bits_raw(port, BIT(pin));
+		}
 	}
 
-	/* set direction */
-	dw_set_bit(port_base_addr, SWPORTA_DDR, pin, (flags & GPIO_DIR_MASK));
-
-	if (interrupt_support && (flags & GPIO_INT)) {
-		dw_interrupt_config(port, GPIO_ACCESS_BY_PIN, pin, flags);
+	/* use built-in debounce */
+	if (dw_interrupt_support(config) && (dir_port == SWPORTA_DDR)) {
+		need_debounce = ((flags & GPIO_INT_ENABLE) != 0U) &&
```

Comment:
You could define `bool need_debounce` here since it's not used outside this block.

But whether or not debounce without interrupt works, you can't trust that the call configuring the pin's non-interrupt features (which currently includes `GPIO_INT_DEBOUNCE`) will have passed `GPIO_INT_ENABLE` (because it's not supposed to in the non-legacy-API case).  So this is still fragile.